### PR TITLE
http Encryption for kerberos and spnego

### PR DESCRIPTION
**Pull request by yakman2020**
Linux->windows.  Linux->Linux worked before, mac->anything is still in progress.
pbuild has been done.
@Microsoft/omi-devs

### CODE REVIEW

File: Unix/http/httpclientauth.c [hunk 1]
```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
+    // We need to get the proper length from the token, but for now 16 bytes is correct for all 
+    // protocols
+
+        signature_length = iov[0].buffer.length;
+
+        output_buffer.length = (iov[0].buffer.length+
+                                iov[1].buffer.length
+                                /* +iov[2].buffer.length+iov[3].buffer.length*/ );
+
+        alloced_data = PAL_Malloc(output_buffer.length); // We retain this so we can tell we alloced it, not gss
```
Comment (paulcallen) [inline Unix/http/httpclientauth.c]: fail operation if this fails!

File: Unix/http/httpclientauth.c [hunk 2]
```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
+            
+             signature_length = 16;  // NTLM_SIGNATURE_LENGTH
+        }
+        maj_stat = (*_g_gssClientState.Gss_Wrap)(&min_stat, handler->authContext, (handler->negoFlags & GSS_C_CONF_FLAG),
+                        GSS_C_QOP_DEFAULT, &input_buffer, &out_flags, &output_buffer);
+        if (maj_stat != GSS_S_COMPLETE)
+        {
+            _ReportError(handler, "gss_wrap failed", maj_stat, min_stat);
+            (*_g_gssClientState.Gss_Release_Buffer)(&min_stat, &output_buffer);
+            return MI_FALSE;
```
Comment (paulcallen) [inline Unix/http/httpclientauth.c]: you will leak the allocated memory. There are other places that leak in failure cases too.

File: Unix/http/httpclientauth.c [hunk 3]
```diff
@@ -297,6 +301,22 @@ static const char GSS_LIBRARY_NAME[] = CONFIG_GSSLIB;
 static const char KRB5_LIBRARY_NAME[] = CONFIG_KRB5LIB;
 
 
+MI_Boolean Gss_Oid_Equal(const gss_OID poid1, const gss_OID poid2) 
+
```
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: Stray blank line here and at 307?

File: Unix/http/httpclientauth.c [hunk 4]
```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
     OM_uint32 min_stat, maj_stat;
     int out_flags;
+    OM_uint32 signature_length = 16;  // NTLM_SIGNATURE_LENGTH
 
-    maj_stat = (*_g_gssClientState.Gss_Wrap)(&min_stat, handler->authContext, (handler->negoFlags & (GSS_C_INTEG_FLAG | GSS_C_CONF_FLAG)),
-                        GSS_C_QOP_DEFAULT, &input_buffer, &out_flags, &output_buffer);
+    char *alloced_data = NULL;  // We will have a leak if we don't free the memory we alloc for the output buffer. 
+
+#if GSS_USE_IOV
+
```
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: Stray blank line? You don't have one at 1476.

File: Unix/http/httpclientauth.c (hunk 1 above)
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: This gets freed at 1672, but there are other code paths to worry about here (I guess Paul pointed out one of these). Please go through ALL possible code paths once this is allocated to be certain it's not leaked down the road!

File: Unix/http/httpclientauth.c [hunk 5]
```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
+        }
+    
+        if (iov[1].buffer.length)
+        {
+            memcpy(outbufp, iov[1].buffer.value, iov[1].buffer.length);
+            outbufp += iov[1].buffer.length;
+        }
+
+        (*_g_gssClientState.Gss_Release_Iov_Buffer)(&min_stat, iov, 2);
+
```
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: Stray blank line since you don't have one at 1410 (or 1459)

File: Unix/http/httpclientauth.c [hunk 6]
```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
+    }
+    else 
+    {
+        if (handler->selectedMech == AUTH_MECH_NTLMSSP)
+        {
+            //  We are here because gss_wrap_iov is not implemented in gss-ntlmssp. 
+            //  If we aren't doing NTLM I'm not sure how we would figure out the signature length
+            
+             signature_length = 16;  // NTLM_SIGNATURE_LENGTH
```
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: I see the comment here (1461-1462). Is the signature length always this? What if it's not, what happens in the code path?

File: Unix/http/httpclientauth.c [hunk 7]
```diff
@@ -1465,13 +1563,16 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
...
     pdst += MULTIPART_ENCRYPTED_LEN;
     memcpy(pdst, original_host_header, host_header_len);
     pdst += host_header_len; 
-    *pdst++ ='\r';
-    *pdst++ ='\n';
+    *pdst++ = '\r';
+    *pdst++ = '\n';
 
+
+//    PAL_Free(*pHeader); We hav no way of knowing how the pHeader got there
```
Comment (jeffaco) [inline Unix/http/httpclientauth.c]: Commented out. Not sure what you're saying here. Is the header leaked? Is the header not leaked? What are you intending to do with pHeader?

File: Unix/http/httpclientauth.c (hunk 1 above)
Comment (yakman2020) [inline Unix/http/httpclientauth.c]: done

File: Unix/http/httpclientauth.c (hunk 6 above)
Comment (yakman2020) [inline Unix/http/httpclientauth.c]: Kerberos has a different siglen, but it is not constant.  NTLM is always 16, so we preinit to 16 and change it if we are doing kerb.

**Comment by yakman2020** [comment]
@Microsoft/omi-devs   Feedback incorporated

**Comment by yakman2020** [comment]
@Microsoft/omi-devs updated

**Pull request merged**